lib/interface_quda.cpp

```diff
@@ -5459,16 +5473,13 @@ void performWFlownStep(unsigned int n_steps, double step_size, int meas_interval
 
   GaugeField *in = gaugeSmeared;
   GaugeField *out = gaugeAux;
-
-  QudaGaugeObservableParam param = newQudaGaugeObservableParam();
-  param.compute_plaquette = QUDA_BOOLEAN_TRUE;
-  param.compute_qcharge = QUDA_BOOLEAN_TRUE;
+  
+  gaugeObservables(*in, param[0], profileWFlow);
```

Comment:
@maddyscientist @cpviolator This change makes the flow itself a little more expensive compared to what was previously there for verbosity levels below `QUDA_SUMMARIZE`. Is this acceptable or do we need to add some more logic?

One could, for example, change lines 5449 and 5450 above to be `getVerbosity() >= QUDA_SUMMARIZE ? QUDA_BOOLEAN_TRUE : QUDA_BOOLEAN_FALSE` to keep the original behaviour of `performWFlownStep`.

Reply:
I think I prefer @kostrzewa's suggestion of setting the booleans according to verbosity level. That will render the call to gaugeObservables a NOP and not waste any FLOPS.

I'm was unsure as to why it is necessary to move the call outside the verbosity boolean guard, but now I see that there an SU(N) projection step in the call to gaugeObservables. Is this the motivation for moving the call?

Reply:
In general, I thought it could be possible that we want to call quda from tmLQCD with  `QUDA_SILENT` but we would like to have the `QudaGaugeObservableParam *param` computed.
I can remove the change if it is in conflict with the logic in quda.

Reply:
@cpviolator In the original implementation, the only way to get QUDA to compute the gauge observables was to switch to a higher verbosity level. Instead, we would like to have QUDA compute the observables for us for a whole Wilson flow integration trajectory via the new `performWFlownStep_param` (when called from tmLQCD) without having lots of console output at the same time *and* without changing the interface of the original `performWFlownStep` function. This seemed like a reasonable way to achieve what we want to do (modulo the change suggested).

If you now say, however, that there is an SU(N) projection hidden in `gaugeObservables`, then we are of course adding operations even if we don't compute the plaquette and topological charge...

Reply:
@Marcogarofalo Ahh, so tmLQCD can pass in the array of param structs, have QUDA be silent, then have tmLQCD dump the information. I think the extra expense of a single call to gaugeObservables is OK.

Reply:
So moving the call outside the guard is actually solving a hidden bug, because verbosity should have no relevance at all to an suN_project policy. The default behaviour in `develop` is to not perform suN projection. I think the addition of `getVerbosity() >= QUDA_SUMMARIZE ? QUDA_BOOLEAN_TRUE : QUDA_BOOLEAN_FALSE` is good.

@maddyscientist What are your thoughts on default behaviour WRT reunitarization during flow computation? We do not have a CL option for it in develop. I would prefer to keep it as (no default reunitarization), add a CL option, and a warning in the flow function giving the state of the reunitarization. This may be why MILC data and QUDA data differ slightly as MILC reunitarises all the time.

Reply:
The cost of computing the plaquette and the topological charge is negligible overall here, so I don't see any reason to just do them anyway.  It's not like this will be called after every solve.

@cpviolator Interesting thought regarding the potential source of the deviation between MILC and QUDA.  Is this something you can look into?

Reply:
@maddyscientist I'll hotfix `develop` with a CL addition for su projection and re run against MILC data to see if there is any effect.

Reply:
I think https://github.com/lattice/quda/pull/1249/commits/2b66ed620e5a6f2a8a4e75a7f9807a055eb02963 should be propagated to all the gauge smearing routines as it is the correct way to call `gaugeObservables` which I will include in the hotfix. This does, however, induce a problem with respect to how reunitarisation is enabled/disabled. Each smearing interface function call creates its own `gaugeObservablesParam` struct. At present there is no way for the user to interact with it. Updating the function signature in `interface_quda.cpp` comes with the usual pitfall of breaking code downstream.

One way to avoid that is to introduce `smearingFunction(Args..., gaugeObservablesParam param = nullptr)` signatures in `quda.h` so that any existing downstream code that uses `smearingFunction(Args...)` remains as is, but users are free to change their code to `smearingFunction(Args..., user_defined_gauge_obs_param)`. Naturally the smearing functions will detect when the `gaugeObservablesParam` arg is a nullptr and dictate its own behaviour.

Any thoughts on this before I go ahead?

Reply:
@kostrzewa @Marcogarofalo Just had an offline conversation with maddyscientist and it's probably better for all if we switch all the smearing routines to behave in this fashion (with arrays of QUDA param structs) and ditch the old interfaces. This will allow for more fine grained control over verbosity and reunitarisation.

If you don't mind I can add the hotfix additions to this branch, and update the testing routines so that arrays of param structs are passed. I'll get this done sooner rather than later so your PR is not held up.

Reply:
thank you, that ok for me.